synapse/storage/databases/main/roommember.py

```diff
@@ -1636,6 +1697,84 @@ def __init__(
             columns=["user_id", "room_id"],
         )
 
+        self.db_pool.updates.register_background_update_handler(
+            "populate_participant_bg_update", self._populate_participant
+        )
+
+    async def _populate_participant(self, progress: JsonDict, batch_size: int) -> int:
+        """
+        Background update to populate column `participant` on `room_memberships` table
+
+         A 'participant' is someone who is currently joined to a room and has sent at least
+        one `m.room.message` or `m.room.encrypted` event.
+        """
+        stream_token = progress.get("last_stream_token", None)
+
+        def _get_max_stream_token_txn(txn: LoggingTransaction) -> int:
+            sql = """
+                SELECT event_stream_ordering from room_memberships
+                ORDER BY event_stream_ordering DESC
+                LIMIT 1;
+            """
+            txn.execute(sql)
+            res = txn.fetchone()
+            if not res or not res[0]:
+                return 0
+            return res[0]
+
+        def _background_populate_participant_txn(
+            txn: LoggingTransaction, stream_token: str
+        ) -> None:
+            sql = """
+                UPDATE room_memberships
+                SET participant = True
+                FROM (
+                    SELECT DISTINCT c.state_key, e.room_id
+                    FROM current_state_events AS c
+                    INNER JOIN events AS e ON c.room_id = e.room_id
+                    WHERE c.membership = 'join'
+                        AND c.state_key = e.sender
+                        AND (
+                            e.type = 'm.room.message'
+                            OR e.type = 'm.room.encrypted'
+                        )
+                ) AS subquery
+                WHERE room_memberships.user_id = subquery.state_key
+                    AND room_memberships.room_id = subquery.room_id
+                    AND room_memberships.event_stream_ordering <= ?
+                    AND room_memberships.event_stream_ordering >= 0;
```

Comment:
With our real-time participant marking code, a given row in the `room_memberships` table having `participant=true` depends on whether the user was a member of the room at the time *and* sent a message while in that session.

This background update will always mark those rows as `participant=false` if the user isn't currently joined to the room. And it will mark `participant=true` if they sent a message *at any point* in the room's history, not just when they were joined that one time.

I think we can eliminate the latter discrepancy by restricting the JOINed `events` table by `stream_ordering` as well. It would also speed up the query considerably.

The former would require fetching the user's membership in the room from somewhere other `current_state_events`, which is trickier. Unless you see a succinct way to do so, I suggest we just leave that one (and add a note about it in the background update method's docstring).

Reply:
```suggestion
                UPDATE room_memberships
                SET participant = True
                FROM (
                    SELECT DISTINCT c.state_key, e.room_id
                    FROM current_state_events AS c
                    INNER JOIN events AS e ON c.room_id = e.room_id
                    WHERE c.membership = 'join'
                        AND c.state_key = e.sender
                        AND (
                            e.type = 'm.room.message'
                            OR e.type = 'm.room.encrypted'
                        )
                ) AS subquery
                WHERE room_memberships.user_id = subquery.state_key
                    AND room_memberships.room_id = subquery.room_id
                    AND room_memberships.event_stream_ordering <= ?
                    AND room_memberships.event_stream_ordering > ?;
```

Where `?` is `stream_token - _POPULATE_PARTICIPANT_BG_UPDATE_BATCH_SIZE`. Otherwise we'll go through the whole historical list in one go.